quarkus/runtime/src/test/java/org/keycloak/quarkus/runtime/services/metrics/events/MicrometerMetricsEventListenerTest.java

```diff
@@ -0,0 +1,790 @@
+
+        final Event event2 = createEvent(EventType.REVOKE_GRANT_ERROR);
+        KeycloakModelUtils.runJobInTransaction(keycloakSessionFactory, (KeycloakSession session) -> {
+            getMicrometerMetricsEventListener(session).onEvent(event2);
+        });
+        assertMetric("keycloak.user", 1, "idp", "",
+                "event", "revoke_grant", "error", "unknown", "client.id", "THE_CLIENT_ID", "realm", DEFAULT_REALM_NAME);
+        assertMetric("keycloak.user", 2, "idp", "",
+                "event", "update_email", "error", "", "client.id", "THE_CLIENT_ID", "realm", DEFAULT_REALM_NAME);
+    }
+
+    @Test
+    public void shouldTolerateNullLabels() {
+        final Event nullEvent = new Event();
+        nullEvent.setType(EventType.LOGIN_ERROR);
+        nullEvent.setClientId(null);
+        nullEvent.setError(null);
+        nullEvent.setRealmId(null);
+
+        KeycloakModelUtils.runJobInTransaction(keycloakSessionFactory, (KeycloakSession session) -> {
+
+            getMicrometerMetricsEventListener(session).onEvent(nullEvent);
+        });
+        assertMetric("keycloak.user", 1, "idp", "",
+                "event", "login", "error", "unknown", "client.id", "", "realm", "");
+    }
+
+    private void assertGenericMetric(String metricName, double metricValue, String... tags) {
+        MatcherAssert.assertThat("Metric value match", meterRegistry.counter(metricName, tags).count() == metricValue);
+    }
+
+    private void assertMetric(String metricName, double metricValue, String... tags) {
+        this.assertGenericMetric(metricName, metricValue, tags);
+    }
+
+    private Event createEvent(EventType type, String realmName, String clientId, String
+            error, Tuple<String, String>... tuples) {
+        final Event event = new Event();
+        event.setType(type);
+        event.setRealmName(realmName);
+        event.setClientId(clientId);
+        if (tuples != null) {
+            event.setDetails(new HashMap<>());
+            for (Tuple<String, String> tuple : tuples) {
+                event.getDetails().put(tuple.left, tuple.right);
+            }
+        } else {
+            event.setDetails(Collections.emptyMap());
+        }
+
+        if (error != null) {
+            event.setError(error);
+        }
+        return event;
+    }
+
+    private Event createEvent(EventType type, String realmName, String
+            clientId, Tuple<String, String>... tuples) {
+        return this.createEvent(type, realmName, clientId, null, tuples);
+    }
+
+    private Event createEvent(EventType type) {
+        return createEvent(type, DEFAULT_REALM_NAME, "THE_CLIENT_ID", (String) null);
+    }
+
+    private static MicrometerUserEventMetricsEventListenerProvider getMicrometerMetricsEventListener(KeycloakSession session) {
+        return session.getProvider(MicrometerUserEventMetricsEventListenerProvider.class);
+    }
+
+    private KeycloakSessionFactory createKeycloakSessionFactory() {
```

Comment:
When possible we are trying to avoid mocks like this. Possibly after moving this test to the base testsuite it would be possible to get rid of this.